pegen/pegen.c

```diff
@@ -153,6 +158,20 @@ CONSTRUCTOR(Parser *p, ...)
     return cache;
 }
 
+static int
+_get_keyword_or_name_type(Parser *p, char *name, size_t name_len)
+{
+    for (int i = 0; i < p->n_keywords; i++) {
+        KeywordToken k = (*p->keywords)[i];
+        if (k.str[0] == name[0]
+            && strlen(k.str) == name_len
+            && strncmp(k.str, name, name_len) == 0) {
```

Comment:
This sounds convoluted. As Pablo mentioned, `strcmp()` is highly optimized (basically the compiler inlines tricks like these) so you're probably better off just writing only `if (strcmp(k.str, name)) {`. (Unless you've timed it already?)

If you want to try a different kind of optimization, maybe you could generate separate arrays for each keyword length, so e.g. if the name has length 1 you know it's not a keyword, if it has length 2 you only compare it to `if`, `is`, `in`, `as`, `or`, and so on.

Reply:
This is not an optimization. This is actually needed, because `name` does not only contain the current token, but everything from the current token until the end of input. This means that `strcmp` would fail, even if the current token is indeed a keyword.

Reply:
OK, got it. I think it's worth a comment. (Also, a future optimization would be to have separate lists of keywords by length.)